course_flow/views.py

```diff
@@ -829,156 +821,24 @@ def dialog_form_create(request: HttpRequest) -> HttpResponse:
-            except ValidationError:
-                return JsonResponse({"action": "error"})
-            return JsonResponse({"action": "posted"})
-    elif model == "preparation":
-        del data["work_classification"], data["activity_classification"]
-        data["parent_preparation"] = None
-        serializer = PreparationSerializer(data=data)
-        if parent_id:
-            week = Week.objects.get(id=parent_id)
-            if serializer.is_valid():
-                preparation = serializer.save()
-            else:
-                return JsonResponse({"action": "error"})
-            try:
-                component = Component.objects.create(
-                    content_object=preparation
-                )
-                for link in ComponentWeek.objects.filter(week=week):
-                    link.rank += 1
-                    link.save()
-                ComponentWeek.objects.create(week=week, component=component)
-            except ValidationError:
-                return JsonResponse({"action": "error"})
-            return JsonResponse({"action": "posted"})
-    elif model == "week":
-        del data["work_classification"], data["activity_classification"]
-        data["parent_week"] = None
-        serializer = WeekSerializer(data=data)
-        if parent_id:
-            course = Course.objects.get(id=parent_id)
-            if serializer.is_valid():
-                week = serializer.save()
-            else:
-                return JsonResponse({"action": "error"})
-            try:
-                for link in WeekCourse.objects.filter(course=course):
-                    link.rank += 1
-                    link.save()
-                WeekCourse.objects.create(course=course, week=week)
-            except ValidationError:
-                return JsonResponse({"action": "error"})
-            return JsonResponse({"action": "posted"})
-    elif model == "course":
+    else:
         del data["work_classification"], data["activity_classification"]
-        data["parent_course"] = None
-        serializer = CourseSerializer(data=data)
-        if parent_id:
-            program = Program.objects.get(id=parent_id)
-            if serializer.is_valid():
-                course = serializer.save()
-            else:
-                return JsonResponse({"action": "error"})
-            try:
-                component = Component.objects.create(content_object=course)
-                for link in ComponentProgram.objects.filter(program=program):
-                    link.rank += 1
-                    link.save()
-                ComponentProgram.objects.create(
-                    program=program, component=component
-                )
-            except ValidationError:
-                return JsonResponse({"action": "error"})
-            return JsonResponse({"action": "posted"})
+        return save_serializer(serializer_lookups[model](data=data))
```

Comment:
This function should always return a JsonResponse

Reply:
That doesn't seem to be the case in master branch. In the cases where model == "program" or when none of the options apply the view returns save_serializer([Model]Serializer(data=data)). When I was going through testing I noticed that this was required in order to pass many of the tests that use this method to create activities, courses, and programs, which I think is why I went back and added this default.

Reply:
Yeah you're right, bad practice on my part I think. I'll have to take a look.